File: YASAudio/Basic/YASAudioFile.m
//
//  YASAudioFile.m
//  Copyright (c) 2015 Yuki Yasoshima.
//

#import "YASAudioFile.h"
#import "YASAudioFormat.h"
#import "YASAudioData.h"
#import "YASMacros.h"
#import "YASAudioTypes.h"
#import "YASAudioUtility.h"
#import "NSException+YASAudio.h"
#import "NSError+YASAudio.h"

NSString *const YASAudioFileType3GPP = @"public.3gpp";
NSString *const YASAudioFileType3GPP2 = @"public.3gpp2";
NSString *const YASAudioFileTypeAIFC = @"public.aifc-audio";
NSString *const YASAudioFileTypeAIFF = @"public.aiff-audio";
NSString *const YASAudioFileTypeAMR = @"org.3gpp.adaptive-multi-rate-audio";
NSString *const YASAudioFileTypeAC3 = @"public.ac3-audio";
NSString *const YASAudioFileTypeMPEGLayer3 = @"public.mp3";
NSString *const YASAudioFileTypeCoreAudioFormat = @"com.apple.coreaudio-format";
NSString *const YASAudioFileTypeMPEG4 = @"public.mpeg-4";
NSString *const YASAudioFileTypeAppleM4A = @"com.apple.m4a-audio";
NSString *const YASAudioFileTypeWAVE = @"com.microsoft.waveform-audio";

#pragma mark - C Utility Function

static Boolean YASOpenExtAudioFileWithFileURL(ExtAudioFileRef *extAudioFile, const NSURL *url)
{
    CFURLRef cfurl = CFBridgingRetain(url);
    OSStatus err = ExtAudioFileOpenURL(cfurl, extAudioFile);
    CFRelease(cfurl);
    return err == noErr;
}

static Boolean YASCreateExtAudioFileWithFileURL(ExtAudioFileRef *extAudioFile, const NSURL *url,
                                                const AudioFileTypeID fileType, const AudioStreamBasicDescription *asbd)
{
    CFURLRef cfurl = CFBridgingRetain(url);
    OSStatus err = ExtAudioFileCreateWithURL(cfurl, fileType, asbd, NULL, kAudioFileFlags_EraseFile, extAudioFile);
    CFRelease(cfurl);
    return err == noErr;
}

static Boolean YASDisposeExtAudioFile(const ExtAudioFileRef extAudioFile)
{
    OSStatus err = ExtAudioFileDispose(extAudioFile);
    return err == noErr;
}

static AudioFileTypeID YASGetAudioFileType(const AudioFileID fileID)
{
    UInt32 fileType;
    UInt32 size = sizeof(AudioFileTypeID);
    YASRaiseIfAUError(AudioFileGetProperty(fileID, kAudioFilePropertyFileFormat, &size, &fileType));
    return fileType;
}

static Boolean YASGetExtAudioFileFormat(AudioStreamBasicDescription *asbd, const ExtAudioFileRef extAudioFile)
{
    UInt32 size = sizeof(AudioStreamBasicDescription);
    OSStatus err = noErr;
    YASRaiseIfAUError(err = ExtAudioFileGetProperty(extAudioFile, kExtAudioFileProperty_FileDataFormat, &size, asbd));
    return err == noErr;
}

static Boolean YASSetClientFormat(const AudioStreamBasicDescription *asbd, const ExtAudioFileRef extAudioFile)
{
    UInt32 size = sizeof(AudioStreamBasicDescription);
    OSStatus err = noErr;
    YASRaiseIfAUError(err = ExtAudioFileSetProperty(extAudioFile, kExtAudioFileProperty_ClientDataFormat, size, asbd));
    return err == noErr;
}

static SInt64 YASGetFileLengthFrames(const ExtAudioFileRef extAudioFile)
{
    SInt64 result = 0;
    UInt32 size = sizeof(SInt64);
    YASRaiseIfAUError(ExtAudioFileGetProperty(extAudioFile, kExtAudioFileProperty_FileLengthFrames, &size, &result));
    return result;
}

static Boolean YASOpenAudioFileWithFileURL(AudioFileID *fileID, const NSURL *url)
{
    CFURLRef cfurl = CFBridgingRetain(url);
    OSStatus err = AudioFileOpenURL(cfurl, kAudioFileReadPermission, kAudioFileWAVEType, fileID);
    CFRelease(cfurl);
    return err == noErr;
}

static Boolean YASCloseAudioFile(const AudioFileID fileID)
{
    OSStatus err = AudioFileClose(fileID);
    return err == noErr;
}

static Boolean YASGetAudioFileFormat(AudioStreamBasicDescription *asbd, const AudioFileID fileID)
{
    UInt32 size = sizeof(AudioStreamBasicDescription);
    OSStatus err = AudioFileGetProperty(fileID, kAudioFilePropertyDataFormat, &size, asbd);
    return err == noErr;
}

static Boolean YASCanOpenAudioFile(const NSURL *url)
{
    Boolean result = true;
    AudioFileID fileID;
    AudioStreamBasicDescription asbd;
    if (YASOpenAudioFileWithFileURL(&fileID, url)) {
        if (!YASGetAudioFileFormat(&asbd, fileID)) {
            result = false;
        }
        YASCloseAudioFile(fileID);
    } else {
        result = false;
    }
    return result;
}

static AudioFileID YASGetAudioFileIDFromExtAudioFile(const ExtAudioFileRef extAudioFile)
{
    UInt32 size = sizeof(AudioFileID);
    AudioFileID audioFileID = 0;
    YASRaiseIfAUError(ExtAudioFileGetProperty(extAudioFile, kExtAudioFileProperty_AudioFile, &size, &audioFileID));
    return audioFileID;
}

static AudioFileTypeID YASAudioFileTypeIDFromFileType(NSString *fileType)
{
    if ([fileType isEqualToString:YASAudioFileType3GPP]) {
        return kAudioFile3GPType;
    } else if ([fileType isEqualToString:YASAudioFileType3GPP2]) {
        return kAudioFile3GP2Type;
    } else if ([fileType isEqualToString:YASAudioFileTypeAIFC]) {
        return kAudioFileAIFCType;
    } else if ([fileType isEqualToString:YASAudioFileTypeAIFF]) {
        return kAudioFileAIFFType;
    } else if ([fileType isEqualToString:YASAudioFileTypeAMR]) {
        return kAudioFileAMRType;
    } else if ([fileType isEqualToString:YASAudioFileTypeAC3]) {
        return kAudioFileAC3Type;
    } else if ([fileType isEqualToString:YASAudioFileTypeMPEGLayer3]) {
        return kAudioFileMP3Type;
    } else if ([fileType isEqualToString:YASAudioFileTypeCoreAudioFormat]) {
        return kAudioFileCAFType;
    } else if ([fileType isEqualToString:YASAudioFileTypeMPEG4]) {
        return kAudioFileMPEG4Type;
    } else if ([fileType isEqualToString:YASAudioFileTypeAppleM4A]) {
        return kAudioFileM4AType;
    } else if ([fileType isEqualToString:YASAudioFileTypeWAVE]) {
        return kAudioFileWAVEType;
    }
    return 0;
}

static NSString *YASFileTypeFromAudioFileTypeID(AudioFileTypeID fileTypeID)
{
    switch (fileTypeID) {
        case kAudioFile3GPType:
            return YASAudioFileType3GPP;
            break;
        case kAudioFile3GP2Type:
            return YASAudioFileType3GPP2;
            break;
        case kAudioFileAIFCType:
            return YASAudioFileTypeAIFC;
            break;
        case kAudioFileAIFFType:
            return YASAudioFileTypeAIFF;
            break;
        case kAudioFileAMRType:
            return YASAudioFileTypeAMR;
            break;
        case kAudioFileAC3Type:
            return YASAudioFileTypeAC3;
            break;
        case kAudioFileMP3Type:
            return YASAudioFileTypeMPEGLayer3;
            break;
        case kAudioFileCAFType:
            return YASAudioFileTypeCoreAudioFormat;
            break;
        case kAudioFileMPEG4Type:
            return YASAudioFileTypeMPEG4;
            break;
        case kAudioFileM4AType:
            return YASAudioFileTypeAppleM4A;
            break;
        case kAudioFileWAVEType:
            return YASAudioFileTypeWAVE;
            break;
        default:
            break;
    }
    return nil;
}

#pragma mark - YASAudioFile

@interface YASAudioFile ()

@property (nonatomic, strong) NSURL *url;
@property (nonatomic, strong) YASAudioFormat *fileFormat;
@property (nonatomic, copy) NSString *fileType;
@property (nonatomic, assign) ExtAudioFileRef extAudioFile;

@end

@implementation YASAudioFile {
    SInt64 _fileFramePosition;
}

- (void)dealloc
{
    [self close];

    YASRelease(_url);
    YASRelease(_fileFormat);
    YASRelease(_processingFormat);
    YASRelease(_fileType);

    _url = nil;
    _fileFormat = nil;
    _processingFormat = nil;
    _fileType = nil;

    YASSuperDealloc;
}

- (SInt64)fileLength
{
    NSAssert(_extAudioFile, nil);
    return YASGetFileLengthFrames(_extAudioFile);
}

- (SInt64)processingLength
{
    const SInt64 fileLength = self.fileLength;
    const Float64 processingPerFileRate =
        _processingFormat.streamDescription->mSampleRate / _fileFormat.streamDescription->mSampleRate;
    return fileLength * processingPerFileRate;
}

- (UInt32)fileFramePosition
{
    return (UInt32)_fileFramePosition;
}

- (void)setFileFramePosition:(UInt32)fileFramePosition
{
    if (_fileFramePosition != fileFramePosition) {
        OSStatus err = ExtAudioFileSeek(_extAudioFile, fileFramePosition);
        if (err == noErr) {
            _fileFramePosition = fileFramePosition;
        }
    }
}

- (void)setProcessingFormat:(YASAudioFormat *)processingFormat
{
    if (processingFormat != _processingFormat) {
        YASRelease(_processingFormat);
        _processingFormat = YASRetain(processingFormat);
        YASSetClientFormat(processingFormat.streamDescription, _extAudioFile);
    }
}

- (SInt64 *)fileFramePositionPointer
{
    return &_fileFramePosition;
}

#pragma mark Private

- (BOOL)_openWithPCMFormat:(YASAudioPCMFormat)pcmFormat interleaved:(BOOL)interleaved
{
    if (!YASCanOpenAudioFile(_url)) {
        return NO;
    }

    if (!YASOpenExtAudioFileWithFileURL(&_extAudioFile, _url)) {
        _extAudioFile = NULL;
        return NO;
    };

    AudioStreamBasicDescription asbd;
    if (!YASGetExtAudioFileFormat(&asbd, _extAudioFile)) {
        [self close];
        return NO;
    }

    AudioFileID audioFileID = YASGetAudioFileIDFromExtAudioFile(_extAudioFile);
    AudioFileTypeID audioFileTypeID = YASGetAudioFileType(audioFileID);
    self.fileType = YASFileTypeFromAudioFileTypeID(audioFileTypeID);
    if (!_fileType) {
        [self close];
        return NO;
    }

    YASAudioFormat *fileFormat = [[YASAudioFormat alloc] initWithStreamDescription:&asbd];
    self.fileFormat = fileFormat;
    YASRelease(fileFormat);

    YASAudioFormat *processingFormat = [[YASAudioFormat alloc] initWithPCMFormat:pcmFormat
                                                                      sampleRate:fileFormat.sampleRate
                                                                        channels:fileFormat.channelCount
                                                                     interleaved:interleaved];
    self.processingFormat = processingFormat;
    YASRelease(processingFormat);

    if (!YASSetClientFormat(_processingFormat.streamDescription, _extAudioFile)) {
        [self close];
        return NO;
    }

    return YES;
}

- (BOOL)_createWithSettings:(NSDictionary *)settings
                  pcmFormat:(YASAudioPCMFormat)pcmFormat
                interleaved:(BOOL)interleaved
{
    YASAudioFormat *fileFormat = [[YASAudioFormat alloc] initWithSettings:settings];
    self.fileFormat = fileFormat;
    YASRelease(fileFormat);

    AudioFileTypeID fileTypeID = YASAudioFileTypeIDFromFileType(_fileType);
    if (!fileTypeID) {
        return NO;
    }

    if (!YASCreateExtAudioFileWithFileURL(&_extAudioFile, _url, fileTypeID, _fileFormat.streamDescription)) {
        _extAudioFile = NULL;
        return NO;
    }

    YASAudioFormat *processingFormat = [[YASAudioFormat alloc] initWithPCMFormat:pcmFormat
                                                                      sampleRate:fileFormat.sampleRate
                                                                        channels:fileFormat.channelCount
                                                                     interleaved:interleaved];
    self.processingFormat = processingFormat;
    YASRelease(processingFormat);

    if (!YASSetClientFormat(_processingFormat.streamDescription, _extAudioFile)) {
        [self close];
        return NO;
    }

    return YES;
}

- (void)close
{
    if (_extAudioFile) {
        YASDisposeExtAudioFile(_extAudioFile);
        _extAudioFile = NULL;
    }
}

#pragma mark - Tests

+ (AudioFileTypeID)_audioFileTypeIDFromFileType:(NSString *)fileType
{
    return YASAudioFileTypeIDFromFileType(fileType);
}

+ (NSString *)_fileTypeFromAudioFileTypeID:(AudioFileTypeID)fileTypeID
{
    return YASFileTypeFromAudioFileTypeID(fileTypeID);
}

@end

#pragma mark - YASAudioFileReader

@implementation YASAudioFileReader

- (instancetype)initWithURL:(NSURL *)fileURL error:(NSError **)outError
{
    return [self initWithURL:fileURL pcmFormat:YASAudioPCMFormatFloat32 interleaved:NO error:outError];
}

- (instancetype)initWithURL:(NSURL *)fileURL
                  pcmFormat:(YASAudioPCMFormat)format
                interleaved:(BOOL)interleaved
                      error:(NSError **)outError
{
    self = [super init];
    if (self) {
        if (!fileURL) {
            if (outError) {
                *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeArgumentIsNil];
            }
            YASRelease(self);
            return nil;
        }
        self.url = fileURL;
        if (![self _openWithPCMFormat:format interleaved:interleaved]) {
            if (outError) {
                *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeNotOpen];
            }
            YASRelease(self);
            return nil;
        }
    }
    return self;
}

- (BOOL)readIntoData:(YASAudioData *)data error:(NSError **)outError
{
    return [self readIntoData:data frameLength:data.frameCapacity error:outError];
}

- (BOOL)readIntoData:(YASAudioData *)data frameLength:(const UInt32)frameLength error:(NSError **)outError
{
    if (![data.format isEqualToAudioFormat:self.processingFormat]) {
        if (outError) {
            *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeInvalidFormat];
        }
        return NO;
    }

    OSStatus err = noErr;
    UInt32 outFrameLength = 0;
    UInt32 remainFrames = frameLength;

    AudioBufferList *ioAbl = YASAudioAllocateAudioBufferListWithoutData(data.format.bufferCount, 0);

    while (remainFrames) {
        UInt32 bytesPerFrame = data.format.streamDescription->mBytesPerFrame;
        UInt32 dataByteSize = remainFrames * bytesPerFrame;
        UInt32 dataIndex = outFrameLength * bytesPerFrame;

        for (NSInteger i = 0; i < data.format.bufferCount; i++) {
            AudioBuffer *audioBuffer = &ioAbl->mBuffers[i];
            audioBuffer->mNumberChannels = data.format.stride;
            audioBuffer->mDataByteSize = dataByteSize;
            audioBuffer->mData = &data.audioBufferList->mBuffers[i].mData[dataIndex];
        }

        UInt32 ioFrames = remainFrames;

        err = ExtAudioFileRead(self.extAudioFile, &ioFrames, ioAbl);
        if (err != noErr) {
            if (outError) {
                *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeReadFailed audioErrorCode:err];
            }
            break;
        }

        if (!ioFrames) {
            break;
        }
        remainFrames -= ioFrames;
        outFrameLength += ioFrames;
    }

    free(ioAbl);

    data.frameLength = outFrameLength;

    if (err == noErr) {
        err = ExtAudioFileTell(self.extAudioFile, self.fileFramePositionPointer);
        if (err != noErr) {
            if (outError) {
                *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeTellFailed audioErrorCode:err];
            }
        }
    }

    return err == noErr;
}

@end

#pragma mark - YASAudioFileWriter

@implementation YASAudioFileWriter

- (instancetype)initWithURL:(NSURL *)fileURL
                   fileType:(NSString *)fileType
                   settings:(NSDictionary *)settings
                      error:(NSError **)outError
{
    return [self initWithURL:fileURL
                    fileType:fileType
                    settings:settings
                   pcmFormat:YASAudioPCMFormatFloat32
                 interleaved:NO
                       error:outError];
}

- (instancetype)initWithURL:(NSURL *)fileURL
                   fileType:(NSString *)fileType
                   settings:(NSDictionary *)settings
                  pcmFormat:(YASAudioPCMFormat)format
                interleaved:(BOOL)interleaved
                      error:(NSError **)outError
{
    self = [super init];
    if (self) {
        if (!fileURL || !fileType || !settings) {
            if (outError) {
                *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeArgumentIsNil];
            }
            YASRelease(self);
            return nil;
        }
        self.url = fileURL;
        self.fileType = fileType;
        if (![self _createWithSettings:settings pcmFormat:format interleaved:interleaved]) {
            if (outError) {
                *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeNotCreate];
            }
            YASRelease(self);
            return nil;
        }
    }
    return self;
}

- (BOOL)writeSyncFromData:(YASAudioData *)data error:(NSError **)outError
{
    return [self _writeFromData:data isAsync:NO error:outError];
}

- (BOOL)writeAsyncFromData:(YASAudioData *)data error:(NSError **)outError
{
    return [self _writeFromData:data isAsync:YES error:outError];
}

- (BOOL)_writeFromData:(YASAudioData *)data isAsync:(BOOL)isAsync error:(NSError **)outError
{
    if (![data.format isEqualToAudioFormat:self.processingFormat]) {
        if (outError) {
            *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeInvalidFormat];
        }
        return NO;
    }

    OSStatus err = noErr;

    if (isAsync) {
        err = ExtAudioFileWriteAsync(self.extAudioFile, data.frameLength, data.audioBufferList);
    } else {
        err = ExtAudioFileWrite(self.extAudioFile, data.frameLength, data.audioBufferList);
    }

    if (err != noErr) {
        if (outError) {
            *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeWriteFailed audioErrorCode:err];
        }
    }

    if (err == noErr) {
        err = ExtAudioFileTell(self.extAudioFile, self.fileFramePositionPointer);
        if (err != noErr) {
            if (outError) {
                *outError = [NSError yas_errorWithCode:YASAudioFileErrorCodeTellFailed audioErrorCode:err];
            }
        }
    }

    return err == noErr;
}

@end

#pragma mark - NSDictionary Category

@implementation NSDictionary (YASAudioFile)

+ (NSDictionary *)yas_waveFileSettingsWithSampleRate:(Float64)sampleRate
                                    numberOfChannels:(UInt32)channels
                                            bitDepth:(UInt32)bitDepth
{
    return [self yas_linearPCMSettingsWithSampleRate:sampleRate
                                    numberOfChannels:channels
                                            bitDepth:bitDepth
                                         isBigEndian:NO
                                             isFloat:bitDepth >= 32
                                    isNonInterleaved:NO];
}

+ (NSDictionary *)yas_aiffFileSettingsWithSampleRate:(Float64)sampleRate
                                    numberOfChannels:(UInt32)channels
                                            bitDepth:(UInt32)bitDepth
{
    return [self yas_linearPCMSettingsWithSampleRate:sampleRate
                                    numberOfChannels:channels
                                            bitDepth:bitDepth
                                         isBigEndian:YES
                                             isFloat:bitDepth >= 32
                                    isNonInterleaved:NO];
}

+ (NSDictionary *)yas_linearPCMSettingsWithSampleRate:(Float64)sampleRate
                                     numberOfChannels:(UInt32)channels
                                             bitDepth:(UInt32)bitDepth
                                          isBigEndian:(BOOL)isBigEndian
                                              isFloat:(BOOL)isFloat
                                     isNonInterleaved:(BOOL)isNonInterleaved
{
    return @{
        AVFormatIDKey: @(kAudioFormatLinearPCM),
        AVSampleRateKey: @(sampleRate),
        AVNumberOfChannelsKey: @(channels),
        AVLinearPCMBitDepthKey: @(bitDepth),
        AVLinearPCMIsBigEndianKey: @(isBigEndian),
        AVLinearPCMIsFloatKey: @(isFloat),
        AVLinearPCMIsNonInterleaved: @(isNonInterleaved),
        AVChannelLayoutKey: [NSData data]
    };
}

+ (NSDictionary *)yas_aacSettingsWithSampleRate:(Float64)sampleRate
                               numberOfChannels:(UInt32)channels
                                       bitDepth:(UInt32)bitDepth
                                encoderQuallity:(AVAudioQuality)encoderQuality
                                 encoderBitRate:(UInt32)bitRate
                            encoderBitDepthHint:(UInt32)bitDepthHint
                     sampleRateConverterQuality:(AVAudioQuality)converterQuality
{
    return @{
        AVFormatIDKey: @(kAudioFormatMPEG4AAC),
        AVSampleRateKey: @(sampleRate),
        AVNumberOfChannelsKey: @(channels),
        AVLinearPCMBitDepthKey: @(bitDepth),
        AVLinearPCMIsBigEndianKey: @(NO),
        AVLinearPCMIsFloatKey: @(NO),
        AVEncoderAudioQualityKey: @(encoderQuality),
        AVEncoderBitRateKey: @(bitRate),
        AVEncoderBitDepthHintKey: @(bitDepthHint),
        AVSampleRateConverterAudioQualityKey: @(converterQuality)
    };
}

@end

File: audio/file/yas_audio_file.cpp
//
//  yas_audio_file.mm
//

#include "yas_audio_file.h"

#include <AudioToolbox/AudioToolbox.h>
#include <cpp_utils/yas_cf_utils.h>
#include <cpp_utils/yas_exception.h>
#include <cpp_utils/yas_fast_each.h>
#include <cpp_utils/yas_result.h>

#include "yas_audio_pcm_buffer.h"

using namespace yas;
using namespace yas::audio;

#pragma mark -

file::file() {
}

file::~file() {
    this->close();
}

file::open_result_t file::open(open_args args) {
    if (this->_ext_audio_file) {
        return open_result_t(open_error_t::opened);
    }

    if (args.pcm_format == pcm_format::other) {
        return open_result_t(open_error_t::invalid_argument);
    }

    this->_path = args.file_path;

    if (!this->_open_ext_audio_file(args.pcm_format, args.interleaved)) {
        return open_result_t(open_error_t::open_failed);
    }

    return open_result_t(nullptr);
}

file::create_result_t file::create(create_args args) {
    if (this->_ext_audio_file) {
        return create_result_t(create_error_t::created);
    }

    if (!args.settings) {
        return create_result_t(create_error_t::invalid_argument);
    }

    this->_path = args.file_path;
    this->_file_type = args.file_type;

    if (!this->_create_ext_audio_file(args.settings, args.pcm_format, args.interleaved)) {
        return create_result_t(create_error_t::create_failed);
    }

    return create_result_t(nullptr);
}

void file::close() {
    if (this->_ext_audio_file) {
        ext_audio_file_utils::dispose(this->_ext_audio_file.value());
        this->_ext_audio_file = std::nullopt;
    }
}

bool file::is_opened() const {
    return this->_ext_audio_file != nullptr;
}

std::filesystem::path const &file::path() const {
    return *this->_path;
}

audio::file_type file::file_type() const {
    return this->_file_type;
}

format const &file::file_format() const {
    return *this->_file_format;
}

format const &file::processing_format() const {
    return *this->_processing_format;
}

int64_t file::file_length() const {
    if (this->_ext_audio_file) {
        return ext_audio_file_utils::get_file_length_frames(this->_ext_audio_file.value());
    }
    return 0;
}

int64_t file::processing_length() const {
    if (!this->_processing_format || !this->_file_format) {
        return 0;
    }

    auto const fileLength = file_length();
    auto const rate = this->_processing_format->stream_description().mSampleRate /
                      this->_file_format->stream_description().mSampleRate;
    return fileLength * rate;
}

int64_t file::file_frame_position() const {
    return this->_file_frame_position;
}

void file::set_processing_format(format format) {
    this->_processing_format = std::move(format);
    if (this->_ext_audio_file) {
        ext_audio_file_utils::set_client_format(this->_processing_format->stream_description(),
                                                this->_ext_audio_file.value());
    }
}

void file::set_file_frame_position(uint32_t const position) {
    if (this->_ext_audio_file && this->_file_frame_position != position) {
        OSStatus err = ExtAudioFileSeek(this->_ext_audio_file.value(), position);
        if (err == noErr) {
            this->_file_frame_position = position;
        }
    }
}

file::read_result_t file::read_into_buffer(pcm_buffer &buffer, uint32_t const frame_length) {
    if (!this->_ext_audio_file) {
        return read_result_t(read_error_t::closed);
    }

    if (buffer.format() != this->_processing_format) {
        return read_result_t(read_error_t::invalid_format);
    }

    if (buffer.frame_capacity() < frame_length) {
        return read_result_t(read_error_t::frame_length_out_of_range);
    }

    OSStatus err = noErr;
    uint32_t out_frame_length = 0;
    uint32_t remain_frames = frame_length > 0 ? frame_length : buffer.frame_capacity();

    auto const &format = buffer.format();
    uint32_t const buffer_count = format.buffer_count();
    uint32_t const stride = format.stride();

    if (auto abl_ptr = allocate_audio_buffer_list(buffer_count, 0, 0).first) {
        AudioBufferList *io_abl = abl_ptr.get();

        while (remain_frames) {
            uint32_t bytesPerFrame = format.stream_description().mBytesPerFrame;
            uint32_t dataByteSize = remain_frames * bytesPerFrame;
            uint32_t dataIndex = out_frame_length * bytesPerFrame;

            auto each = make_fast_each(buffer_count);
            while (yas_each_next(each)) {
                auto const &idx = yas_each_index(each);
                AudioBuffer *ab = &io_abl->mBuffers[idx];
                ab->mNumberChannels = stride;
                ab->mDataByteSize = dataByteSize;
                uint8_t *byte_data = static_cast<uint8_t *>(buffer.audio_buffer_list()->mBuffers[idx].mData);
                ab->mData = &byte_data[dataIndex];
            }

            UInt32 io_frames = remain_frames;

            err = ExtAudioFileRead(this->_ext_audio_file.value(), &io_frames, io_abl);
            if (err != noErr) {
                break;
            }

            if (!io_frames) {
                break;
            }
            remain_frames -= io_frames;
            out_frame_length += io_frames;
        }
    }

    buffer.set_frame_length(out_frame_length);

    if (err != noErr) {
        return read_result_t(read_error_t::read_failed);
    } else {
        err = ExtAudioFileTell(this->_ext_audio_file.value(), &this->_file_frame_position);
        if (err != noErr) {
            return read_result_t(read_error_t::tell_failed);
        }
    }

    return read_result_t(nullptr);
}

file::write_result_t file::write_from_buffer(pcm_buffer const &buffer, bool const async) {
    if (!this->_ext_audio_file) {
        return write_result_t(write_error_t::closed);
    }

    ExtAudioFileRef const &ext_audio_file = this->_ext_audio_file.value();

    if (buffer.format() != this->_processing_format) {
        return write_result_t(write_error_t::invalid_format);
    }

    OSStatus err = noErr;

    if (async) {
        err = ExtAudioFileWriteAsync(ext_audio_file, buffer.frame_length(), buffer.audio_buffer_list());
    } else {
        err = ExtAudioFileWrite(ext_audio_file, buffer.frame_length(), buffer.audio_buffer_list());
    }

    if (err != noErr) {
        return write_result_t(write_error_t::write_failed);
    } else {
        err = ExtAudioFileTell(ext_audio_file, &this->_file_frame_position);
        if (err != noErr) {
            return write_result_t(write_error_t::tell_failed);
        }
    }

    return write_result_t(nullptr);
}

#pragma mark - private

bool file::_open_ext_audio_file(pcm_format const pcm_format, bool const interleaved) {
    if (!ext_audio_file_utils::can_open(this->_path.value())) {
        return false;
    }

    ExtAudioFileRef ext_audio_file = nullptr;
    if (!ext_audio_file_utils::open(&ext_audio_file, this->_path.value())) {
        this->_ext_audio_file = std::nullopt;
        return false;
    };
    this->_ext_audio_file = ext_audio_file;

    AudioStreamBasicDescription asbd;
    if (!ext_audio_file_utils::get_audio_file_format(&asbd, ext_audio_file)) {
        this->close();
        return false;
    }

    AudioFileTypeID file_type_id = ext_audio_file_utils::get_audio_file_type_id(ext_audio_file);

    try {
        this->_file_type = to_file_type(file_type_id);
    } catch (std::exception const &) {
        this->close();
        return false;
    }

    this->_file_format = format{asbd};

    this->_processing_format = format{{.sample_rate = _file_format->sample_rate(),
                                       .channel_count = this->_file_format->channel_count(),
                                       .pcm_format = pcm_format,
                                       .interleaved = interleaved}};

    if (!ext_audio_file_utils::set_client_format(this->_processing_format->stream_description(), ext_audio_file)) {
        this->close();
        return false;
    }

    return true;
}

bool file::_create_ext_audio_file(CFDictionaryRef const &settings, pcm_format const pcm_format,
                                  bool const interleaved) {
    this->_file_format = format{settings};

    AudioFileTypeID file_type_id = to_audio_file_type_id(this->_file_type);

    ExtAudioFileRef ext_audio_file = nullptr;
    if (!ext_audio_file_utils::create(&ext_audio_file, this->_path.value(), file_type_id,
                                      this->_file_format->stream_description())) {
        this->_ext_audio_file = std::nullopt;
        return false;
    }
    this->_ext_audio_file = ext_audio_file;

    this->_processing_format = format{{.sample_rate = this->_file_format->sample_rate(),
                                       .channel_count = this->_file_format->channel_count(),
                                       .pcm_format = pcm_format,
                                       .interleaved = interleaved}};

    if (!ext_audio_file_utils::set_client_format(this->_processing_format->stream_description(), ext_audio_file)) {
        this->close();
        return false;
    }

    return true;
}

#pragma mark -

file_ptr file::make_shared() {
    return file_ptr(new file{});
}

file::make_opened_result_t file::make_opened(file::open_args args) {
    auto file = make_shared();
    if (auto result = file->open(std::move(args))) {
        return file::make_opened_result_t{std::move(file)};
    } else {
        return file::make_opened_result_t{std::move(result.error())};
    }
}

file::make_created_result_t file::make_created(file::create_args args) {
    auto file = make_shared();
    if (auto result = file->create(std::move(args))) {
        return file::make_created_result_t{std::move(file)};
    } else {
        return file::make_created_result_t{std::move(result.error())};
    }
}

std::string yas::to_string(file::open_error_t const &error_t) {
    switch (error_t) {
        case file::open_error_t::opened:
            return "opened";
        case file::open_error_t::invalid_argument:
            return "invalid_argument";
        case file::open_error_t::open_failed:
            return "open_failed";
    }
}

std::string yas::to_string(file::read_error_t const &error_t) {
    switch (error_t) {
        case file::read_error_t::closed:
            return "closed";
        case file::read_error_t::invalid_format:
            return "invalid_format";
        case file::read_error_t::read_failed:
            return "read_failed";
        case file::read_error_t::tell_failed:
            return "tell_failed";
        case file::read_error_t::frame_length_out_of_range:
            return "frame_length_out_of_range";
    }
}

std::string yas::to_string(file::create_error_t const &error_t) {
    switch (error_t) {
        case file::create_error_t::created:
            return "created";
        case file::create_error_t::invalid_argument:
            return "invalid_argument";
        case file::create_error_t::create_failed:
            return "create_failed";
    }
}

std::string yas::to_string(file::write_error_t const &error_t) {
    switch (error_t) {
        case file::write_error_t::closed:
            return "closed";
        case file::write_error_t::invalid_format:
            return "invalid_format";
        case file::write_error_t::write_failed:
            return "write_failed";
        case file::write_error_t::tell_failed:
            return "tell_failed";
    }
}

std::ostream &operator<<(std::ostream &os, file::open_error_t const &value) {
    os << to_string(value);
    return os;
}

std::ostream &operator<<(std::ostream &os, file::read_error_t const &value) {
    os << to_string(value);
    return os;
}

std::ostream &operator<<(std::ostream &os, file::create_error_t const &value) {
    os << to_string(value);
    return os;
}

std::ostream &operator<<(std::ostream &os, file::write_error_t const &value) {
    os << to_string(value);
    return os;
}
